assignments/burmanm/week-2.scm

```diff
@@ -0,0 +1,394 @@
+;; 100037
+;; 100039
+;; 100041
+;; 100043 *** 0.
+;; ;Value: 3
+
+;; 1 ]=> (find-next-three 1000000)
+
+;1000001
+;1000003 *** 9.999999999999787e-3
+;1000005
+;1000007
+;1000009
+;1000011
+;1000013
+;1000015
+;1000017
+;1000019
+;1000021
+;1000023
+;1000025
+;1000027
+;1000029
+;1000031
+;1000033 *** 0.
+;1000035
+;1000037 *** 1.0000000000001563e-2
+;Value: 3
+
+
+;; Since times were too fast, using values 1e9, 1e10 and 1e11 were used instead. Those lead us to
+;; around 0.07, 0.21 and 0.66 on my machine. Which is around ~3 times and thus close to sqrt(10).
+
+;; Exercise 1.23
+
+(define (next n)
+  (if (= n 2)
+      3
+      (+ n 2)))
+
+(define (smallest-divisor n) (find-divisor n 2))
+(define (divides? a b) (= (remainder b a) 0))
+(define (find-divisor n test-divisor)
+  (cond ((> (square test-divisor) n) n)
+	((divides? test-divisor n) test-divisor)
+	(else (find-divisor n (next test-divisor)))))
+
+;; With changes:
+(find-next-three 1e11)
+
+;; 100000000001.
+;; 100000000003. *** .39000000000000057
+;; 100000000005.
+;; 100000000007.
+;; 100000000009.
+;; 100000000011.
+;; 100000000013.
+;; 100000000015.
+;; 100000000017.
+;; 100000000019. *** .370000000000001
+;; 100000000021.
+;; 100000000023.
+;; 100000000025.
+;; 100000000027.
+;; 100000000029.
+;; 100000000031.
+;; 100000000033.
+;; 100000000035.
+;; 100000000037.
+;; 100000000039.
+;; 100000000041.
+;; 100000000043.
+;; 100000000045.
+;; 100000000047.
+;; 100000000049.
+;; 100000000051.
+;; 100000000053.
+;; 100000000055.
+;; 100000000057. *** .35999999999999943
+;; ;Value: 3
+
+;; That is, time is around half the previous.
+
+;; Exercise 1.24
+(define (square x)
+   (* x x))
+
+(define (even? n)
+   (= (remainder n 2) 0))
+
+(define (expmod base exp m)
+  (cond ((= exp 0) 1)
+	((even? exp)
+	 (remainder
+	  (square (expmod base (/ exp 2) m))
+	  m))
+	(else
+	 (remainder
+	  (* base (expmod base (- exp 1) m))
+	  m))))
+
+(define (fermat-test n)
+  (define (try-it a)
+    (= (expmod a n n) a))
+  (try-it (+ 1 (random (- n 1)))))
+(define (fast-prime? n times)
+  (cond ((= times 0) true)
+	((fermat-test n) (fast-prime? n (- times 1)))
+	(else false)))
+
+;; Use our modified version of timed-prime-test to reduce printing to only primes
+
+(define (timed-prime-test n)
+  (start-prime-test n (runtime)))
+(define (start-prime-test n start-time)
+  (if (fast-prime? n 100)
+      (report-prime (- (runtime) start-time) n)
+      #f))
+(define (report-prime elapsed-time n)
+  (newline)
+  (display n)
+  (display " *** ")
+  (display elapsed-time)
+  #t)
+
+;; Use (find-next-three (round->exact n)) to get it working with large numbers..
+
+;; The time to calculate fermat numbers does not seem to increase at all and is less than the precision of
+;; the runtime. Replacing runtime with real-time-clock (mit-scheme) gives us a better indication:
+
+(define (timed-prime-test n)
+  (start-prime-test n (real-time-clock)))
+(define (start-prime-test n start-time)
+  (if (fast-prime? n 100)
```

Comment:
I'm surprised there is no indication for times parameter to use. It seems like it significantly affect the time growth of the process.

Reply:
I found this part had particular badly defined exercises with lots of "freedom" to choose how one is going to implement it. Including how to use certain given code (I read multiple versions by other people and noticed that each one had selected different modifications to the original code to get things done).